pkg/scheduler/plugins/gpushare/device_handler.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gpushare
+
+import (
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+)
+
+func (n *nodeDeviceCache) onDeviceAdd(obj interface{}) {
+	device, ok := obj.(*schedulingv1alpha1.Device)
+	if !ok {
+		return
+	}
+	n.assign(device.Name, device)
+	klog.V(4).InfoS("device cache added", "Device", klog.KObj(device))
+}
+
+func (n *nodeDeviceCache) onDeviceUpdate(oldObj, newObj interface{}) {
+	oldD, oldOK := oldObj.(*schedulingv1alpha1.Device)
+	newD, newOK := newObj.(*schedulingv1alpha1.Device)
+	if !oldOK || !newOK {
+		return
+	}
+	if oldD == nil || newD == nil {
+		return
+	}
+	n.assign(newD.Name, newD)
+	klog.V(4).InfoS("device cache updated", "Device", klog.KObj(newD))
+}
+
+func (n *nodeDeviceCache) onDeviceDelete(obj interface{}) {
+	var device *schedulingv1alpha1.Device
+	switch t := obj.(type) {
+	case *schedulingv1alpha1.Device:
+		device = t
+	case cache.DeletedFinalStateUnknown:
+		var ok bool
+		device, ok = t.Obj.(*schedulingv1alpha1.Device)
+		if !ok {
+			return
+		}
+	default:
+		return
+	}
+	n.delete(device.Name)
+	klog.V(4).InfoS("device cache deleted", "Device", klog.KObj(device))
+}
+
+func (n *nodeDeviceCache) delete(nodeName string) {
+	if nodeName == "" {
+		return
+	}
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	delete(n.nodeDeviceInfos, nodeName)
+}
+
+func (n *nodeDeviceCache) assign(nodeName string, device *schedulingv1alpha1.Device) {
+	if nodeName == "" || device == nil {
+		return
+	}
+
+	info := n.getNodeDevice(nodeName)
+	if info == nil {
+		info = newNodeDevice()
+		n.nodeDeviceInfos[nodeName] = info
```

Comment:
"n.nodeDeviceInfos[nodeName] = info" is not protected by n.lock.  "get not exsit then create" is a successive actions that under lock protection. if break lock,  there will be multi-thread problem,  thread A get not exist, thread B get not exist concurrently, then A and B will both try newNodeDevice, which is unexpected.  althrough assign of one node can't be execute by informer really, but is a good habit.